File: src/ViewModels/Sakuno.ING.ViewModels.Logging/LogsVM.vb
Imports Sakuno.ING.Game
Imports Sakuno.ING.Game.Logger

Public MustInherit Class LogsVM(Of T As {Class, ITimedEntity})
    Protected MustOverride Function CreateFilters() As FilterVM(Of T)()
    Protected MustOverride Function GetEntities() As IReadOnlyCollection(Of T)

    Private ReadOnly _filters As IBindableCollection(Of FilterVM(Of T))
    Public ReadOnly Property Filters As IBindableCollection(Of IFilterVM)
        Get
            Return _filters
        End Get
    End Property

    Private ReadOnly _snapshot As New BindableSnapshotCollection(Of T)
    Public ReadOnly Property Entities As IBindableCollection(Of T)
        Get
            Return _snapshot
        End Get
    End Property

    Public Sub New()
        _filters = CreateFilters().ToBindable()
    End Sub

    Public Sub Refresh()
        Dim source = GetEntities().AsEnumerable()
        For Each filter In _filters
            filter.UpdateCandidates(source)
            source = source.Where(AddressOf filter.Hits)
        Next
        _snapshot.Query = source
    End Sub
End Class

File: src/ViewModels/Sakuno.ING.ViewModels.Logging/LogsVM.cs
using System.Collections.Generic;
using System.Linq;
using Sakuno.ING.Game;
using Sakuno.ING.Game.Logger;

namespace Sakuno.ING.ViewModels.Logging
{
    public abstract class LogsVM<T>
        where T : class, ITimedEntity
    {
        private protected abstract FilterVM<T>[] CreateFilters();
        private protected abstract IReadOnlyCollection<T> GetEntities();

        private readonly IBindableCollection<FilterVM<T>> _filters;
        public IBindableCollection<IFilterVM> Filters => _filters;

        private readonly BindableSnapshotCollection<T> snapshot = new BindableSnapshotCollection<T>();
        public IBindableCollection<T> Entities => snapshot;

        public LogsVM()
        {
            _filters = CreateFilters().ToBindable();
            foreach (var filter in _filters)
                filter.Updated += Refresh;
        }

        public void Refresh()
        {
            IEnumerable<T> source = GetEntities();
            foreach(var filter in _filters)
            {
                filter.UpdateCandidates(source);
                if (filter.IsEnabled)
                    source = source.Where(filter.Hits);
            }
            snapshot.Query = source;
        }
    }
}
